File: solutions/php/strain/1/Strain.php
<?php

declare(strict_types=1);

class Strain
{
    public function keep(array $list, callable $predicate): array
    {
        return array_values(array_filter($list, function($v) use ($predicate) {
            return !$predicate || $predicate($v) === true;
        }));
    }

    public function discard(array $list, callable $predicate): array
    {
        return array_values(array_filter($list, function($v) use ($predicate) {
            return !$predicate || $predicate($v) === false;
        }));
    }
}

File: solutions/erlang/strain/1/src/strain.erl
-module(strain).
-export([keep/2, discard/2, test_version/0]).
keep(Fn, List) ->
	filter(Fn, List, true).
discard(Fn, List) ->
	filter(Fn, List, false).
filter(Fn, List, Expect) ->
	filter(Fn, List, Expect, []).
filter(_, [], _, Acc) ->
	lists:reverse(Acc);
filter(Fn, [X|More], Expect, Acc) ->
	case Fn(X) of
		R when R=:=Expect -> filter(Fn, More, Expect, [X|Acc]);
		R when R=:=not Expect -> filter(Fn, More, Expect, Acc)
	end.
test_version() -> 1.